source/common/network/connection_impl.cc

```diff
@@ -194,6 +194,15 @@ Connection::State ConnectionImpl::state() const {
 
 void ConnectionImpl::closeConnectionImmediately() { closeSocket(ConnectionEvent::LocalClose); }
 
+void ConnectionImpl::setTransportSocketIsReadable() {
+  transport_wants_read_ = true;
+  // Only schedule a read activation if read is enabled. If read was disabled, read resumption will
+  // happen once read is re-enabled since transport_wants_read_ is now set to true.
+  if (read_disable_count_ == 0) {
```

Comment:
So I see a test change to ensure that when read_disabled we don't activate here.
there's one other change in this PR, which is that in filterChainWantsData  we no longer set transport_wants_read_ true.  Is that behavior regression tested as well?

Reply:
It stroke me odd as well so I dug a bit deeper into that and I think the current and the previous code are equivalent:

In the previous code we set `transport_wants_read_` via calling  `setReadBufferReady` in `ReadDisable` if:

 (*) `transport_wants_read_` is true already or `buffer.length() > 0`. This comes from the if-statement above. `if (filterChainWantsData() && (read_buffer_.length() > 0 || transport_wants_read_))`

Looking at the two cases individually:

1. In the case that `transport_wants_read_` is true already then it doesn't really matter that we do not set  `transport_wants_read_` to true.
2. In the case that `buffer.length() > 0` then the next call to `ReadDisable(false)` will trigger the read activation regardless of the value of `transport_wants_read_` because of (*).

This means that even in the case where that we have a cycle of ReadDisable(true)/ReadEnable(false) (this used to leave connections at zombie state because the read activation was canceled) we will re-activate reads correctly.

+1 on testing though, the more the merrier.